### DISCUSSION TITLE
Empty collections

### DISCUSSION THREAD

**Discussion by IgorErin** [YaccConstructor/GraphBLAS-sharp#74] [Ideas]
As the basic operations are implemented, support for empty collections (matrices, vectors) becomes more and more necessary.

This problem occurs not only as a result of the main operation, but also when calculating the intermediate result. An example is `Matrix.map`, `SpGeMM` and `Kronecker`, respectively

At the same time, the solution to this problem is not so obvious.

The previously proposed support for the empty format has its semantic drawbacks. 

- Since if an empty matrix is a format such as `CSR`, then therefore another matrix can be converted into it. Or create an empty matrix from various data.

- The usual rotation of the matrix in `Option` does not solve the problem, since it does not allow you to save the size of the matrices.

**Comment by gsvgit** [comment]
What about the fallowing desing?
```
type Matrix<'Storage>
   memeber this.Height with get ...
   memeber this.Width with get ...
  // storage-access-related methods
```

This way any matrix (collection)  should provide common storage-format-independent propertise (size). At the same time any matrix (collection) parametrized by storage and provides acces to it.

**Reply by IgorErin** [reply]
As far as I understand, such a design does not fit well with the current structure of the project. Since it does not allow passing common fields (such as size, context) when working with more specialized abstractions.

For example, when passing a `CSR`-type matrix to a [module](https://github.com/YaccConstructor/GraphBLAS-sharp/blob/e0485c52128b16b0ba4668655b1c9dfc1b033a05/src/GraphBLAS-sharp.Backend/Matrix/CSRMatrix/Matrix.fs#L14) operating with `CSR`-type matrices, you will need to pass the size every time, same for result return.
